src/lib/holocene/api-pagination.svelte

```diff
@@ -1,73 +1,127 @@
 
-  let store = createPaginationStore();
-  let error: any;
+  export let pageSizeOptions: string[] = options;
+  export let total: string | number = '';
 
-  $: nextIndex = $store.nextIndex;
-  $: pageSize = $store.pageSize;
+  let store = createPaginationStore(pageSizeOptions);
+  let error: any;
+  let shifted = false;
 
   function clearError() {
     if (error) error = undefined;
   }
 
-  async function onPageChange(nextIndex: number) {
-    clearError();
-    store.setUpdating();
-    try {
-      const fetchData: PaginatedRequest = await onFetch();
-      const response = await fetchData(pageSize, $store.indexTokens[nextIndex]);
-      const { items, nextPageToken } = response;
-      store.setNextPageToken(nextPageToken, items);
-    } catch (err: any) {
-      error = err;
-      onError(error);
-    }
+  onMount(() => {
+    initalDataFetch();
+  });
+
+  async function initalDataFetch() {
+    const fetchData: PaginatedRequest = await onFetch();
+    const response = await fetchData($store.pageSize, '');
+    const { items, nextPageToken } = response;
+    store.nextPageWithItems(nextPageToken, items);
   }
 
-  async function onPageSizeChange(pageSize: number) {
+  async function fetchIndexData() {
     clearError();
-    store.resetPageSize();
-    try {
-      const fetchData: PaginatedRequest = await onFetch();
-      const response = await fetchData(
-        pageSize,
-        $store.indexTokens[$store.index],
-      );
-      const { items, nextPageToken } = response;
-      store.setNextPageToken(nextPageToken, items);
-    } catch (err: any) {
-      error = err;
-      onError(error);
+    store.setUpdating();
+    if (!$store.hasNextIndexData) {
+      try {
+        const fetchData: PaginatedRequest = await onFetch();
+        const response = await fetchData(
+          $store.pageSize,
+          $store.indexData[$store.index].nextToken,
+        );
+        const { items, nextPageToken } = response;
+        store.nextPageWithItems(nextPageToken, items);
+      } catch (err: any) {
+        error = err;
+        onError(error);
+      }
+    } else {
+      store.nextPage();
     }
   }
 
-  $: {
-    onPageChange(nextIndex);
+  async function handleKeydown(event: KeyboardEvent) {
+    switch (event.code) {
+      case 'ArrowRight':
+        if ($store.hasNext && !$store.updating) {
+          fetchIndexData();
+        }
+        break;
+      case 'ArrowLeft':
+        if ($store.hasPrevious && !$store.updating) {
+          store.previousPage();
+        }
+        break;
+      case 'ArrowUp':
+        if (shifted && onShiftUp) {
+          onShiftUp(event);
+          store.reset();
+          initalDataFetch();
+        } else {
+          store.previousRow();
+        }
+        break;
+      case 'ArrowDown':
+        if (shifted && onShiftDown) {
+          onShiftDown(event);
+          store.reset();
+          initalDataFetch();
+        } else {
+          store.nextRow();
+        }
+        break;
```

Comment:
Saw a nice comment during the demo about using vim keys, i.e. `h`, `j`, `k`, and `l`, maybe worth adding those?

Reply:
Good call, easy to add that.

Reply:
Oh also, I think you can git rid of `shifted` and just use `event.shiftKey` https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/shiftKey

Reply:
Yup that absolutely works and is so much better.